From: Tom Lendacky <thomas.lendacky@amd.com>
Date: Mon, 18 Dec 2017 12:53:04 +0100
Subject: KVM: x86: Add speculative control CPUID support for guests
Patch-mainline: submitted on 2018/1/9
References: bsc#1068032

Provide the guest with the speculative control CPUID related values.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Tom Lendacky <thomas.lendacky@amd.com>
Signed-off-by: Borislav Petkov <bp@suse.de>

---
 arch/x86/kvm/x86.c |   20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -2550,6 +2550,8 @@ static void do_cpuid_ent(struct kvm_cpui
 		F(BMI1) | F(HLE) | F(AVX2) | F(SMEP) | F(BMI2) | F(RTM) | F(FSGSBASE) | F(ERMS) |
 		f_invpcid;
 
+	const u32 kvm_cpuid_7_0_edx_x86_features = F(SPEC_CTRL);
+
 	/* all calls to cpuid_count() should be made on the same cpu */
 	get_cpu();
 	do_cpuid_1_ent(entry, function, index);
@@ -2613,7 +2615,7 @@ static void do_cpuid_ent(struct kvm_cpui
 			entry->ebx = 0;
 		entry->eax = 0;
 		entry->ecx = 0;
-		entry->edx = 0;
+		entry->edx &= kvm_cpuid_7_0_edx_x86_features;
 		break;
 	}
 	/* function 0xb has additional index. */
@@ -2675,6 +2677,22 @@ static void do_cpuid_ent(struct kvm_cpui
 		entry->ecx &= kvm_supported_word6_x86_features;
 		cpuid_mask(&entry->ecx, 6);
 		break;
+
+	case 0x80000008:
+		entry->eax = 0;
+		entry->ecx = 0;
+		entry->edx = 0;
+
+		/*
+		 * cpuid 0x80000008.0.ebx
+		 *
+		 * Boris: hardcode due to prior kABI fix.
+		 */
+		if (boot_cpu_has(X86_FEATURE_IBPB))
+			 entry->ebx |= (1 << 12);
+
+		break;
+
 	/*Add support for Centaur's CPUID instruction*/
 	case 0xC0000000:
 		/*Just support up to 0xC0000004 now*/
